minimap.impl.user.js

```diff
@@ -418,22 +440,109 @@ const { html, render } = mlp_uhtml;
+        })
+        .catch(err => {
+          console.error("Error updating template", err);
+        });
+    // Also update mask if needed
+    if (typeof rPlaceTemplate.maskUrl !== "undefined") {
+      fetchTemplate(rPlaceTemplate.maskUrl)
+          .then(array => {
+            const img = new Image();
+            img.src = createPngDataUrlForBytes(array);
+            img.onload = () => {
+              maskCtx.clearRect(0, 0, maskCanvas.width, maskCanvas.height);
+              maskCtx.drawImage(img, 0, 0);
+              loadMask();
+            };
+          })
+          .catch(err => {
+            console.error("Error updating mask", err);
+          });
+    } else {
+      // Free memory if we don't need it.
+      rPlaceMask = undefined;
+    }
   };
   setInterval(updateTemplate, 1 * 60 * 1000);
   updateTemplate();
 
+  function loadMask() {
+    const maskData = maskCtx.getImageData(0, 0, maskCanvas.width, maskCanvas.height).data;
+
+    rPlaceMask = new Array(maskCanvas.width * maskCanvas.height);
+    for (let i = 0; i < rPlaceMask.length; i++) {
+      // Grayscale, pick green channel!
+      rPlaceMask[i] = maskData[i * 4 + 1];
+    }
+  }
+
+  /**
+   * Pick a bucket, in a weighted way.
+   *
+   * @param {Map<number, any>} buckets
+   * @return {[number, any]}
+   */
+  function pickBucket(buckets) {
+    const bucketSum = Array.from(buckets.keys()).reduce((sum, key) => sum + key, 0);
+    let random = Math.random();
+    let offset = 0;
+    for (const [value, coords] of buckets) {
+      if (random <= (offset + value) / bucketSum) {
+        return [value, coords];
+      }
+      offset += value;
+    }
+    // If for some reason this breaks, just return a random pixel from the largest bucket
+    const value = Array.from(buckets.keys()).reduce((a, b) => Math.max(a, b), 0);
+    return [value, buckets.get(value)];
+  }
+
+  /**
+   * Select a random pixel weighted by the mask.
+   * @param {[number, number][]} diff
+   * @return {[number, number]}
+   */
+  function selectRandomPixelWeighted(diff) {
+    const buckets = new Map();
+    for (let i = 0; i < diff.length; i++) {
+      const coords = diff[i];
+      const [x, y] = coords;
+      const maskValue = rPlaceMask[x + y * rPlaceCanvas.width];
+      if (maskValue === 0) {
+        continue;
+      }
+      const bucket = buckets.get(maskValue);
+      if (bucket === undefined) {
+        buckets.set(maskValue, [coords]);
+      } else {
+        bucket.push(coords);
+      }
+    }
+    console.log("Picking from buckets", buckets);
+    const [value, bucket] = pickBucket(buckets);
+    console.log("Picked the", value, "bucket")
+    return bucket[Math.floor(Math.random() * bucket.length)];
+  }
+
+  /**
+   * Select a random pixel.
+   *
+   * @param {[number, number][]} diff
+   * @return {[number, number]}
+   */
+  function selectRandomPixel(diff) {
+    if (typeof rPlaceTemplate.maskUrl === "undefined" || typeof rPlaceMask === "undefined") {
```

Comment:
nit: `rPlaceTemplate.maskUrl === undefined || rPlaceMask === undefined`

JavaScript equality operators can't be overridden, so this is preferrable.